spectrum-offchain-lm/src/bundle/rocksdb.rs

```diff
@@ -1,60 +1,74 @@
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use rocksdb::{Direction, IteratorMode};
+use tokio::task::spawn_blocking;
+
 use spectrum_offchain::{
     binary::prefixed_key,
     data::{
-        unique_entity::{Confirmed, Predicted, Traced, Unconfirmed},
+        unique_entity::{Confirmed, Predicted, Traced},
         OnChainEntity,
     },
 };
-use tokio::task::spawn_blocking;
 
+use crate::data::bundle::IndexedStakingBundle;
 use crate::data::{AsBox, BundleId, BundleStateId, PoolId};
 
-use super::{BundleRepo, EpochHeightRange, StakingBundle};
+use super::{BundleRepo, StakingBundle};
 
 pub struct BundleRepoRocksDB {
     pub db: Arc<rocksdb::OptimisticTransactionDB>,
     pub epoch_len: u32,
 }
 
-#[async_trait]
-impl BundleRepo for BundleRepoRocksDB {
-    fn get_epoch_length(&self) -> u32 {
-        self.epoch_len
-    }
+fn epoch_index_prefix(pool_id: PoolId, epoch_ix: u32) -> Vec<u8> {
+    let mut prefix_bytes = bincode::serialize(POOL_EPOCH_PREFIX).unwrap();
+    let pool_id_bytes = bincode::serialize(&pool_id).unwrap();
+    let neg_epoch_ix = u32::MAX - epoch_ix;
+    let epoch_ix_bytes = bincode::serialize(&neg_epoch_ix).unwrap();
+    prefix_bytes.extend_from_slice(&pool_id_bytes);
+    prefix_bytes.extend_from_slice(&epoch_ix_bytes);
+    prefix_bytes
+}
 
-    fn get_height_range(&self, epoch_ix: u32) -> EpochHeightRange {
-        let first = (epoch_ix - 1) * self.epoch_len;
-        let last = epoch_ix * self.epoch_len - 1;
-        EpochHeightRange { first, last }
+fn destruct_epoch_index_key(key: &[u8]) -> Option<(BundleId, u32)> {
```

Comment:
You mean `destructure_epoch_index_key` like `let (x:xs) = [1, 2, 3]` in Haskell right?